datalake-api/src/main/java/com/sequenceiq/sdx/api/model/SdxInternalClusterRequest.java

```diff
@@ -2,7 +2,7 @@
 
 import com.sequenceiq.cloudbreak.api.endpoint.v4.stacks.request.StackV4Request;
 
-public class SdxInternalClusterRequest extends SdxClusterRequest {
+public class SdxInternalClusterRequest extends SdxCustomClusterRequest {
```

Comment:
I wonder if it has any benefit (or dare I say, if it makes sense) to allow an extra image UUID override for the internal request. In particular, `SdxInternalClusterRequest` already contains a complete `StackV4Request` that may or may not have an image UUID inside. In any case, I do not think it is a good idea to override the ID inside the `StackV4Request` with the one coming from SdxCustomClusterRequest.

Reply:
if ((stackRequest.getImage() == null) && (sdxClusterRequest instanceof SdxCustomClusterRequest)) {
            stackRequest.setImage(((SdxCustomClusterRequest) sdxClusterRequest).getImageSettingsV4Request());
}

You can see the overwrite happens only if the StackV4Request contains no image settings. Are you still concerned? Peter wants to set image when user specifies it in internal request. I am trying to satisfy both opinions

Reply:
Yes, I have seen the above logic but I think it shall not touch the image ID in `SdxInternalClusterRequest` whatsoever. If the user sending the req omitted the image ID, he/she probably wants to leave it that way. If he/she wanted to have an image ID explicitly, that could have been included in `SdxInternalClusterRequest.stackV4Request`, without relying on the UUID override of `SdxCustomClusterRequest`.

In short, my feeling is `SdxInternalClusterRequest` shall keep `SdxClusterRequest` as its superclass and not have the extras of `SdxCustomClusterRequest`. The logic you quoted will still work in both cases, but it will override the `null image` only when using `SdxCustomClusterRequest` instead of `SdxInternalClusterRequest`.

@sodre90 @doktoric @lnardai Please do not be shy and share your thoughts. 😃

Reply:
Actually, thinking a bit more I would even go further: I think `SdxCustomClusterRequest` should be a type not related to either `SdxClusterRequest` or `SdxInternalClusterRequest`. The cleanest approach would be to make `SdxCustomClusterRequest` be a subclass of `Object`, having all the fields of `SdxClusterRequest` except for the `runtime`. That way the conflict between `runtime` and `SdxCustomClusterRequest.imageSettingsV4Request` could be avoided entirely.

Reply:
@lajosrodek I prefer to not overwrite `image` in `SdxCustomClusterRequest.` In this way, user has full control of its value. Adding additional logic internally could be confusing to user. @sodre90 Are you OK with this? User who issues internal CLI can setting the image, why not let user get what they set directly?

@lajosrodek In the internal CLI, we have the same issue of conflict between `SdxInternalClusterRequest.StackV4Request.runtime` and `SdxInternalClusterRequest.ImageSettingsV4Request`. 

`The cleanest approach would be to make SdxCustomClusterRequest be a subclass of Object, having all the fields of SdxClusterRequest except for the runtime`
How about the following approach?
1. I define SdxCustomClusterRequest as you mentioned
2. then In SdxCustomController, I create `SdxClusterRequest` whose `runtime` is null and add a new input `ImageSettingsV4Request` for SdxService.createSdx(). In this way, we can avoid the conflict totally, and still minimize the code changes in many places.

Reply:
@LinaAtAustin Yes, `SdxInternalClusterRequest` should have also used `Object` as its superclass instead of `SdxClusterRequest`. The `SdxClusterRequest.runtime` is in conflict with `SdxInternalClusterRequest.StackV4Request` because the latter may prescribe a totally different runtime & image combination. We cannot help that here easily.

Instead, what I suggested are
1. Not to make `SdxInternalClusterRequest` be a subclass of `SdxCustomClusterRequest`; and
2. Make `SdxCustomClusterRequest` be a subclass of `Object`.

The controller change (i.e. converting `SdxCustomClusterRequest` into `SdxClusterRequest` and passing the additional `ImageSettingsV4Request` to `SdxService.createSdx()` sounds good. 👍  

Let us wait for some more comments as the rest of the team has not yet expressed its opinions.

Reply:
@LinaAtAustin We would like to avoid any "instanceof". As I see we have too many parameters that can be null, so I understand your point and your code, but as of now, I would recommend if we just add the ImageSettingsV4Request as a new parameter to createSdx method. It is ugly, but it is the easiest way for now and I will create a ticket for us to refactor / rearrange createSdx method.

I think we should create an abstract ancestor class for SdxClusterRequest like SdxClusterRequestBase and put all its parameters from SdxClusterRequest except runtime. SdxClusterRequest and SdxCustomClusterRequest should extend SdxClusterRequestBase. So now we have to add one more parameter to createSdx, the runtime field, and we have to change sdxClusterRequest paramater's class to SdxClusterRequestBase. I admit it is way more ugly now, we added 2 more parameters to createSdx method but on the API side we are fine. Maybe we can create a parameter class with parameters of createSdx but it is not a big improvement :) (userCrn, name, sdxClusterRequestBase, runtime, internalStackV4Request, imageSettingsV4Request).

You can implement your code without instanceOf, but with some null checks (our instanceOf hater soul will be in peace) and we will be fine on API side (no runtime field in custom request). The code is very ugly at the moment (the original one), it will be a little uglier but we can explode this method later, I will create a jira about this because I don't want to push this work to you. We were the ones who created this mess, it is our task to clean it.

I wouldn't touch SdxInternalClusterRequest (I think it shouldn't have runtime anyway, but if we remove it then we will break the API compatibility..)

Reply:
Good point, using a common ancestor for `SdxClusterRequest` and `SdxCustomClusterRequest` to extract common fields makes more sense. 😄

Reply:
I agree. Will update soon